File: 3. Longest Substring Without Repeating Characters/Java/Longest Substring Without Repeating Characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> memory = new HashMap<>();
        int result = 0;
        for (int i = 0, j = 0; i < s.length(); i++) {
            char letter = s.charAt(i);
            if (memory.containsKey(letter)) {
                j = Math.max(j, memory.get(letter) + 1);
            }
            memory.put(letter, i);
            result = Math.max(result, i - j + 1);
        }
        return result;
    }
}

File: 3. Longest Substring Without Repeating Characters/Python/Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        memory = {}
        result = 0
        j = 0
        for i, letter in enumerate(s):
            if letter in memory:
                j = max(j, memory[letter] + 1)
            memory[letter] = i
            result = max(result, i - j + 1)
        return result
